### ISSUE TITLE
`--target-os` doesn't strip sufficiently

### ISSUE THREAD

**Issue by richardwhiuk** [RustSec/cargo-audit#96]
Running `cargo audit --target-os linux` on https://github.com/slog-rs/term suggests that it's vulnerable to `RUSTSEC-2019-0011` on Linux.

Running `cargo tree --all-targets -p memoffset:0.2.1 -i` shows that the only path is as follows:

```
memoffset v0.2.1
└── crossbeam-epoch v0.6.1
    └── crossbeam v0.5.0
        └── rust-argon2 v0.5.0
            └── redox_users v0.3.1
                └── dirs v1.0.5
                    └── term v0.5.2
                        └── slog-term v2.4.1 (/src/term)
```

`dirs` only uses `redox_users` when `cfg(target_os = "redox")` - see https://github.com/soc/dirs-rs/blob/3c3b61ff9611762bece3fc66fd6612b125819e3f/Cargo.toml#L15-L16 

It looks like `--target-os linux` isn't stripping this dependency, despite being specified on the command line.

**Comment by tarcieri** [comment]
The `--target-os` flag filters by advisories which explicitly self-identify as belonging to a particular OS.

Doing what you're describing would require examining the specific target platform-based gating in a set of transitive dependencies, or would require call graph analysis to determine you're not impacted (e.g. #89).

Personally I think the latter is the best approach.

**Comment by richardwhiuk** [comment]
Hmm, I agree the later approach is likely preferable - although the former is likely to be much simpler (and the call graph analysis will need to consider platform triples as well - which might not be the default).

**Comment by tarcieri** [comment]
> the former is likely to be much simpler

Do you know of a crate that can parse this information out of `Cargo.lock` files today? I don't, short of something like trying to link to `cargo` itself, which would massively increase the build times.

**Comment by richardwhiuk** [comment]
Ah, I assumed that cargo audit used the cargo crate to parse the Cargo.lock file....

**Comment by tarcieri** [comment]
Nope, it consumes the list of dependencies computed by Cargo in `Cargo.lock`, using its own (extremely simple) type for parsing it:

https://github.com/RustSec/rustsec-crate/blob/master/src/lockfile.rs#L10

Unfortunately if a dependency is platform-gated in Cargo.toml, but not marked as `optional = true`, it will appear in Cargo.lock regardless of the current platform.

Perhaps we should investigate if we can better integrate with Cargo, e.g. as a Cargo plugin:

https://www.ncameron.org/blog/cargo-in-2019/

/cc @nrc

**Comment by Diggsey** [comment]
This would be really helpful: there are a lot of wasm-specific dependencies that pull in low-level crates (like bump allocators) that tend to have more vulnerabilities, and I really don't care because I'm not running on wasm.
